WIP: Lazy Pagination

NinekoTheCat opened the pull request:
Fixes: #90 by adding a new type of paginator

## Changes
- Added LazyPages and tests for it.
- Added `Pages` as an interface to replace the static `Pages` class input into Paginators.
- Added custom rule to avoid prefixing interfaces with I
## TODOS:
- [ ] Modify paginators to actually recognize and handle endless pages
- [ ] Add more options for Pages types to support previous/next hasPrevious/hasNext

NinekoTheCat committed:
Added LazyPages
Added LazyPages for people to be able to add pages that are lazily loaded in as they're needed.
Changed the Pages API to use an interface for further
customisation.

Signed-off-by: Niko Dale <cnotsomark@gmail.com>

 Author:    Niko Dale <cnotsomark@gmail.com>

gdude2002 commented:
I understand that this is a draft, but just so you're aware, the project convention is to avoid `IName` interfaces, instead preferring a `Name` interface with a `DefaultName` implementation where a default implementation makes sense.

In this case, I believe that means a `Pages` interface, with `DefaultPages` and `LazyPages` implementations, right?

Otherwise, looking good so far.

NinekoTheCat committed:
Fixed license dates for LazyPages Files.

Signed-off-by: Niko Dale <cnotsomark@gmail.com>

NinekoTheCat committed:
Renamed Pages and IPages
Had to be renamed because of repository conventions
- Pages renamed to DefaultPages
- IPages renamed to Pages

Signed-off-by: Niko Dale <cnotsomark@gmail.com>

NinekoTheCat committed:
Added a detekt rule to enforce interface naming.
- updated detekt
- added a new module dev-detekt-rules
- modified .editorconfig to include rule with no wildcard imports.
Signed-off-by: Niko Dale <cnotsomark@gmail.com>

Signed-off-by: Niko Dale <cnotsomark@gmail.com>

NinekoTheCat commented:
> I understand that this is a draft, but just so you're aware, the project convention is to avoid `IName` interfaces, instead preferring a `Name` interface with a `DefaultName` implementation where a default implementation makes sense.
> 
> In this case, I believe that means a `Pages` interface, with `DefaultPages` and `LazyPages` implementations, right?
> 
> Otherwise, looking good so far.

I renamed them to the proper names

NinekoTheCat committed:
Added MutablePages for Pages that can be mutated.
Signed-off-by: Niko Dale <cnotsomark@gmail.com>

Signed-off-by: Niko Dale <cnotsomark@gmail.com>

NinekoTheCat committed:
Changed how LazyPages works to support providers with no size
Signed-off-by: Niko Dale <cnotsomark@gmail.com>

Signed-off-by: Niko Dale <cnotsomark@gmail.com>

NinekoTheCat committed:
Added annotation for the experimental pagination API so that if we release it, it'll have to be explicitly opt-in
Signed-off-by: Niko Dale <cnotsomark@gmail.com>

Signed-off-by: Niko Dale <cnotsomark@gmail.com>
